packages/cardano-services/src/Program/utils.ts

```diff
@@ -0,0 +1,250 @@
+          pool.query(args, values).catch(async (error) => {
+            if (error.code && ['ENOTFOUND', 'ECONNREFUSED'].includes(error.code)) {
+              const address = await dnsResolver(host!);
+              pool = new Pool({ database, host: address.name, password, port: address.port, user });
+              return await pool.query(args, values);
+            }
+            throw error;
+          });
+      }
+      // Bind if it is a function, no intercept operations
+      if (typeof pool[prop as keyof Pool] === 'function') {
+        const method = pool[prop as keyof Pool] as any;
+        return method.bind(pool);
+      }
+    }
+  });
+};
+
+export const getPool = async (dnsResolver: DnsResolver, options?: HttpServerOptions): Promise<Pool | undefined> => {
+  if (options?.dbConnectionString && options.postgresSrvServiceName)
+    throw new InvalidArgsCombination(ProgramOptionDescriptions.DbConnection, ProgramOptionDescriptions.PostgresSrvArgs);
+  if (options?.dbConnectionString) return new Pool({ connectionString: options.dbConnectionString });
+  if (options?.postgresSrvServiceName && options?.postgresUser && options.postgresDb && options.postgresPassword) {
+    return getPoolWithServiceDiscovery(dnsResolver, {
+      database: options.postgresDb,
+      host: options.postgresSrvServiceName,
+      password: options.postgresPassword,
+      user: options.postgresUser
+    });
+  }
+  // If db connection string is not passed nor postgres srv service name
+  return undefined;
+};
+
+export const isOgmiosConnectionError = (error: any) => {
+  if (
+    (error?.innerError?.code && ['ENOTFOUND', 'ECONNREFUSED'].includes(error.innerError.code)) ||
+    error instanceof WebSocketClosed
+  ) {
+    return true;
+  }
+  return false;
+};
+
+/**
+ * Creates a extended TxSubmitProvider instance :
+ * - use passed srv service name in order to resolve the port
+ * - make dealing with failovers (re-resolving the port) opaque
+ * - use exponential backoff retry internally with default timeout and factor
+ * - intercept 'submitTx' operation and handle connection errors runtime
+ * - all other operations are bind to pool object withoud modifications
+ *
+ * @returns TxSubmitProvider instance
+ */
+export const ogmiosTxSubmitProviderWithDiscovery = async (
+  dnsResolver: DnsResolver,
+  serviceName: string
+): Promise<TxSubmitProvider> => {
+  const { name, port } = await dnsResolver(serviceName!);
+  let ogmiosProvider = ogmiosTxSubmitProvider({ host: name, port });
+
+  return new Proxy<TxSubmitProvider>({} as TxSubmitProvider, {
+    get(_, prop) {
+      if (prop === 'then') return;
+      if (prop === 'submitTx') {
+        return (args: Uint8Array) =>
+          ogmiosProvider.submitTx(args).catch(async (error) => {
+            if (isOgmiosConnectionError(error)) {
+              const record = await dnsResolver(serviceName!);
+              ogmiosProvider = ogmiosTxSubmitProvider({ host: record.name, port: record.port });
```

Comment:
We should attempt to close the old provider, same with rabbitmq provider
```suggestion
              await ogmiosProvider.close().catch(() => void 0); // maybe log a message if failed to close
              ogmiosProvider = ogmiosTxSubmitProvider({ host: record.name, port: record.port });
```

Reply:
`Provider` also has a `start?` but that doesn't seem to be used on any provider. Should we remove it from the interface? @rhyslbw @iccicci

Reply:
Let's leave it as-is, then consider it as part of the decorator refactor

Reply:
8fd7c5267f8250848f954350f58bfef3735aea12